Signal/src/contact/OWSContactsManager.m

```diff
@@ -131,34 +131,70 @@ - (void)handleAddressBookChanged
 
 #pragma mark - Setup
 
-- (void)setupAddressBook {
+- (void)setupAddressBookIfNecessary
+{
     dispatch_async(ADDRESSBOOK_QUEUE, ^{
-      [[OWSContactsManager asyncGetAddressBook] thenDo:^(id addressBook) {
-        self.addressBookReference = addressBook;
-        ABAddressBookRef cfAddressBook = (__bridge ABAddressBookRef)addressBook;
-        ABAddressBookRegisterExternalChangeCallback(cfAddressBook, onAddressBookChanged, (__bridge void *)self);
-        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-            [self handleAddressBookChanged];
-        });
-      }];
+        // De-bounce address book setup.
+        if (self.isContactsUpdateInFlight) {
+            return;
+        }
+        // We only need to set up our address book once;
+        // after that we only need to respond to onAddressBookChanged.
+        if (self.addressBookReference) {
+            return;
+        }
+        self.isContactsUpdateInFlight = YES;
+
+        TOCFuture *future = [OWSContactsManager asyncGetAddressBook];
+        [future thenDo:^(id addressBook) {
+            // Success.
+            OWSAssert(self.isContactsUpdateInFlight);
+            OWSAssert(!self.addressBookReference);
+
+            self.addressBookReference = addressBook;
+            self.isContactsUpdateInFlight = NO;
+
+            ABAddressBookRef cfAddressBook = (__bridge ABAddressBookRef)addressBook;
+            ABAddressBookRegisterExternalChangeCallback(cfAddressBook, onAddressBookChanged, (__bridge void *)self);
+            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                [self handleAddressBookChanged];
+            });
+        }];
+        [future catchDo:^(id failure) {
+            // Failure.
+            OWSAssert(self.isContactsUpdateInFlight);
+            OWSAssert(!self.addressBookReference);
+
+            self.isContactsUpdateInFlight = NO;
+        }];
     });
 }
 
 - (void)intersectContacts {
+    [self intersectContactsWithRetryDelay:60];
+}
+
+- (void)intersectContactsWithRetryDelay:(CGFloat)retryDelaySeconds
+{
+    void (^success)() = ^{
+        DDLogInfo(@"%@ Successfully intersected contacts.", self.tag);
+        [self fireSignalRecipientsDidChange];
+    };
+    void (^failure)(NSError *error) = ^(NSError *error) {
+        DDLogWarn(@"%@ Failed to intersect contacts with error: %@. Rescheduling", self.tag, error);
+
+        // Retry with exponential backoff.
+        //
+        // TODO: Abort if another contact
+        // intersection succeeds in the meantime.
+        dispatch_after(
```

Comment:
Only failure we were just hammering the service with retries, even if we hit the rate limit.  Hmm, I should modify this to abort if we get a status code that indicates the rate limit has been hit.